### PR TITLE
Fix scene load crash related to `_ready`

### LINKED ISSUE #76890: Crash when opening a specific scene
### Godot version

4.1 dev2

### System information

Windows 10 x64

### Issue description

```
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.1.dev.custom_build (554a1e2485abe728346ed9f239cf67a7e0e394a3)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] Node::_propagate_ready (C:\godot_source\scene\main\node.cpp:177)
[1] Node::_propagate_ready (C:\godot_source\scene\main\node.cpp:181)
[2] Node::_set_tree (C:\godot_source\scene\main\node.cpp:2628)
[3] Node::_add_child_nocheck (C:\godot_source\scene\main\node.cpp:1143)
[4] Node::add_child (C:\godot_source\scene\main\node.cpp:1160)
[5] EditorNode::set_edited_scene (C:\godot_source\editor\editor_node.cpp:3455)
[6] EditorNode::load_scene (C:\godot_source\editor\editor_node.cpp:3753)
[7] EditorNode::open_request (C:\godot_source\editor\editor_node.cpp:3937)
[8] EditorNode::_menu_option_confirm (C:\godot_source\editor\editor_node.cpp:2560)
[9] EditorNode::_menu_option (C:\godot_source\editor\editor_node.cpp:1419)
[10] call_with_variant_args_helper<EditorNode,int,0> (C:\godot_source\core\variant\binder_common.h:308)
[11] call_with_variant_args<EditorNode,int> (C:\godot_source\core\variant\binder_common.h:418)
[12] CallableCustomMethodPointer<EditorNode,int>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[13] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[14] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1065)
[15] Object::emit_signal<int> (C:\godot_source\core\object\object.h:868)
[16] PopupMenu::activate_item (C:\godot_source\scene\gui\popup_menu.cpp:1822)
[17] PopupMenu::activate_item_by_event (C:\godot_source\scene\gui\popup_menu.cpp:1744)
[18] MenuBar::shortcut_input (C:\godot_source\scene\gui\menu_bar.cpp:166)
[19] Node::_call_shortcut_input (C:\godot_source\scene\main\node.cpp:2815)
[20] SceneTree::_call_input_pause (C:\godot_source\scene\main\scene_tree.cpp:957)
[21] Viewport::push_unhandled_input (C:\godot_source\scene\main\viewport.cpp:2943)
[22] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:2914)
[23] Window::_window_input (C:\godot_source\scene\main\window.cpp:1387)
[24] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:303)
[25] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:418)
[26] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[27] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[28] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:2297)
[29] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:2262)
[30] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:706)
[31] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:964)
[32] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:1991)
[33] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1475)
[34] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:181)
[35] _main (C:\godot_source\platform\windows\godot_windows.cpp:203)
[36] main (C:\godot_source\platform\windows\godot_windows.cpp:217)
[37] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:231)
[38] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[39] <couldn't map PC to fn name>
-- END OF BACKTRACE --
```
Not sure what causes this, but I managed to isolate it. It's related to one of my scripts, but the stack trace doesn't really point to anything specific.

It doesn't crash in 4.0.2

### Steps to reproduce

1. Open the MRP
2. Open CrashHere.tscn
3. If it doesn't crash, close the scene and open again (it doesn't always happen)

### Minimal reproduction project

[YAC.zip](https://github.com/godotengine/godot/files/11436916/YAC.zip)
The crash seems to be caused by the NPC script.

### LINKED ISSUE #77610: Not all children nodes receive _ready() call
### Godot version

4.1 dev

### System information

Win 10

### Issue description

Have a node and bunch of child nodes.
Some get _ready called, and some do not.

### Steps to reproduce

create a scene and bunch of child nodes with different sripts.

### Minimal reproduction project

n/a

### PR DESCRIPTION
**Pull request by Sauermann**
The iterator might get invalidated during `_ready`-user-code by using `set_name()`. 
Prevent this by disallowing `set_name` during iterator-operations.

resolve #76890
resolve #77610

### CODE REVIEW

**Comment by akien-mga** [comment]
Would this happen to solve #77610 too? This issue just came on my radar, seems related.

File: scene/main/node.cpp [hunk 1]
```diff
@@ -217,8 +217,13 @@ void Node::_notification(int p_notification) {
 void Node::_propagate_ready() {
 	data.ready_notified = true;
 	data.blocked++;
+
+	List<Node *> li;
```
Comment (RedworkDE) [inline scene/main/node.cpp:221]:
```suggestion
	TightLocalVector<Node *> li;
	li.reserve(data.children.size());
```
`List` is a linked list, it is almost always better to use a continious container instead.

**Comment by Sauermann** [comment]
> Would this happen to solve #77610 too?

After testing the MRP of that bugreport, I can confirm, that this PR fixes it too.

File: scene/main/node.cpp (hunk 1 above)
Comment (Sauermann) [reply scene/main/node.cpp]: Thanks for the improvement. I will try to remember this method.

**Comment by akien-mga** [comment]
This looks to me like a good solution, which should be safe.

The regressions were likely introduced by https://github.com/godotengine/godot/pull/75627, which greatly optimized the performance of adding/removing nodes.

We might want to do some new performance tests with this PR to see the impact of splitting this operation in two loops with an intermediate data structure. It might have a significant impact.

**Comment by Sauermann** [comment]
I see this PR only as a temporary solution for the most prominent case of the problem, that the user can rename nodes during `_ready`. This solution should be superseded by a change of the `Node::data.children` data-structure.
I believe, that the actual problem is, that the user has write-access to the Keys of `Node::data.children`.

Looking at `node.cpp`, there are multiple locations like
``` c++
	if (!K.value->is_inside_tree()) { // could have been added in enter_tree
		K.value->_propagate_enter_tree();
	}
```
where the user has the chance to influence `data.children` by renaming nodes and has the chance to cause crashes.

So I suggest instead:
`HashMap<int, Node *> children;`
with internal integer-keys, that are unique per Node.

**Comment by RedworkDE** [comment]
Not sure what exactly your idea is, but adding / changing to such a map probably kills performance in other places instead.

A more limited change would be to add a method to hash map that can replace the key of an element in-place, without reallocating the internal element so that the iterator isnt broken by a rename operation.

**Comment by Sauermann** [comment]
> A more limited change would be to add a method to hash map that can replace the key of an element in-place, without reallocating the internal element so that the iterator isnt broken by a rename operation.

That sounds like a very interesting idea.

**Comment by Sauermann** [comment]
I have updated the PR and removed the creation of a new list of all children.
Instead I utilize `data.children_cache` for the iteration.

**Comment by AThousandShips** [comment]
Might be good to do the same for `propagate_call` and I think a few other places where the map is used like that

**Comment by AThousandShips** [comment]
An alternative interim solution would be to block things that alters the hash map when ever it's used, to be followed up by with a cherrypick when a solution closer to this has been tested, at least if this gets difficult to verify

**Comment by Sauermann** [comment]
This should (untested) also safeguard against iterator-invalidation, but it looks too restrictive, that renaming Nodes becomes impossible in these situations.
```diff
diff --git a/scene/main/node.cpp b/scene/main/node.cpp
index a21866610c..b90273d1ff 100644
--- a/scene/main/node.cpp
+++ b/scene/main/node.cpp
@@ -457,9 +457,11 @@ void Node::_propagate_groups_dirty() {
 		}
 	}
 
+	data.blocked++;
 	for (KeyValue<StringName, Node *> &K : data.children) {
 		K.value->_propagate_groups_dirty();
 	}
+	data.blocked--;
 }
 
 void Node::add_child_notify(Node *p_child) {
@@ -595,9 +597,11 @@ void Node::_propagate_pause_notification(bool p_enable) {
 		notification(NOTIFICATION_UNPAUSED);
 	}
 
+	data.blocked++;
 	for (KeyValue<StringName, Node *> &K : data.children) {
 		K.value->_propagate_pause_notification(p_enable);
 	}
+	data.blocked--;
 }
 
 Node::ProcessMode Node::get_process_mode() const {
@@ -615,12 +619,14 @@ void Node::_propagate_process_owner(Node *p_owner, int p_pause_notification, int
 		notification(p_enabled_notification);
 	}
 
+	data.blocked++;
 	for (KeyValue<StringName, Node *> &K : data.children) {
 		Node *c = K.value;
 		if (c->data.process_mode == PROCESS_MODE_INHERIT) {
 			c->_propagate_process_owner(p_owner, p_pause_notification, p_enabled_notification);
 		}
 	}
+	data.blocked--;
 }
 
 void Node::set_multiplayer_authority(int p_peer_id, bool p_recursive) {
@@ -628,9 +634,11 @@ void Node::set_multiplayer_authority(int p_peer_id, bool p_recursive) {
 	data.multiplayer_authority = p_peer_id;
 
 	if (p_recursive) {
+		data.blocked++;
 		for (KeyValue<StringName, Node *> &K : data.children) {
 			K.value->set_multiplayer_authority(p_peer_id, true);
 		}
+		data.blocked--;
 	}
 }
 
@@ -887,6 +895,7 @@ void Node::_remove_tree_from_process_thread_group() {
 		return; // May not be initialized yet.
 	}
 
+	data.blocked++;
 	for (KeyValue<StringName, Node *> &K : data.children) {
 		if (K.value->data.process_thread_group != PROCESS_THREAD_GROUP_INHERIT) {
 			continue;
@@ -894,6 +903,7 @@ void Node::_remove_tree_from_process_thread_group() {
 
 		K.value->_remove_tree_from_process_thread_group();
 	}
+	data.blocked--;
 
 	if (_is_any_processing()) {
 		_remove_from_process_thread_group();
@@ -912,6 +922,7 @@ void Node::_add_tree_to_process_thread_group(Node *p_owner) {
 		data.process_group = &data.tree->default_process_group;
 	}
 
+	data.blocked++;
 	for (KeyValue<StringName, Node *> &K : data.children) {
 		if (K.value->data.process_thread_group != PROCESS_THREAD_GROUP_INHERIT) {
 			continue;
@@ -919,6 +930,7 @@ void Node::_add_tree_to_process_thread_group(Node *p_owner) {
 
 		K.value->_add_to_process_thread_group();
 	}
+	data.blocked--;
 }
 bool Node::is_processing_internal() const {
 	return data.process_internal;
@@ -1133,6 +1145,7 @@ void Node::_set_name_nocheck(const StringName &p_name) {
 
 void Node::set_name(const String &p_name) {
 	ERR_FAIL_COND_MSG(data.inside_tree && !Thread::is_main_thread(), "Changing the name to nodes inside the SceneTree is only allowed from the main thread. Use call_deferred(\"set_name\",new_name).");
+	ERR_FAIL_COND_MSG(data.blocked > 0, "Parent node is busy setting up children, `set_name()` failed. Consider using `set_name.call_deferred()` instead.");
 	String name = p_name.validate_node_name();
 
 	ERR_FAIL_COND(name.is_empty());
@@ -1811,9 +1824,11 @@ void Node::get_owned_by(Node *p_by, List<Node *> *p_owned) {
 		p_owned->push_back(this);
 	}
 
+	data.blocked++;
 	for (KeyValue<StringName, Node *> &K : data.children) {
 		K.value->get_owned_by(p_by, p_owned);
 	}
+	data.blocked--;
 }
 
 void Node::_set_owner_nocheck(Node *p_owner) {
@@ -3005,9 +3020,11 @@ void Node::get_argument_options(const StringName &p_function, int p_idx, List<St
 
 void Node::clear_internal_tree_resource_paths() {
 	clear_internal_resource_paths();
+	data.blocked++;
 	for (KeyValue<StringName, Node *> &K : data.children) {
 		K.value->clear_internal_tree_resource_paths();
 	}
+	data.blocked--;
 }
 
 PackedStringArray Node::get_configuration_warnings() const {
```

**Comment by AThousandShips** [comment]
I think the block is only required where some method is called that can be overridden, don't know about `get_owned_by` for example

File: scene/main/node.cpp [hunk 2]
```diff
@@ -217,8 +217,10 @@ void Node::_notification(int p_notification) {
 void Node::_propagate_ready() {
 	data.ready_notified = true;
 	data.blocked++;
-	for (KeyValue<StringName, Node *> &K : data.children) {
-		K.value->_propagate_ready();
+
+	_update_children_cache();
```
Comment (reduz) [inline scene/main/node.cpp:221]: While the fix is valid, this has severe performance consequences so it should not be made.
Instead, I suggest just checking for `data.blocked > 0`  during set_name() and suggest the user to use call_deferred("set_name"). instead.

Comment (reduz) [reply scene/main/node.cpp:221]: As a rule, calling _update_children_cache() can be very expensive, so it should only be done when accessing children by indices.

Comment (Sauermann) [reply scene/main/node.cpp]: Thanks for the insight. I have updated the PR according to your suggestions.

File: scene/main/node.cpp [hunk 3]
```diff
@@ -1133,6 +1137,7 @@ void Node::_set_name_nocheck(const StringName &p_name) {
 
 void Node::set_name(const String &p_name) {
 	ERR_FAIL_COND_MSG(data.inside_tree && !Thread::is_main_thread(), "Changing the name to nodes inside the SceneTree is only allowed from the main thread. Use call_deferred(\"set_name\",new_name).");
```
Comment (kleonc) [inline scene/main/node.cpp:1139]:
```suggestion
	ERR_FAIL_COND_MSG(data.inside_tree && !Thread::is_main_thread(), "Changing the name to nodes inside the SceneTree is only allowed from the main thread. Use `set_name.call_deferred(new_name)`.");
```

Comment (Sauermann) [reply scene/main/node.cpp]: Thanks, i have updated the PR with this change.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**